File: archive-esp-idf/main/plc_runtime.c
#include "plc_runtime.h"

#include <string.h>

#include "archive_io.h"
#include "esp_log.h"
#include "nvs.h"
#include "nvs_flash.h"

static const char *TAG = "archive_plc";
static const char *NVS_NAMESPACE = "archive_plc";
static const char *NVS_KEY_RETAIN = "retain";

static void set_safe_outputs(archive_runtime_t *runtime)
{
    runtime->outputs.drive_run_cmd = false;
    runtime->outputs.alarm_light = false;
    runtime->outputs.run_light = false;
}

static esp_err_t load_retain(archive_retain_t *retain)
{
    nvs_handle_t handle;
    esp_err_t err = nvs_open(NVS_NAMESPACE, NVS_READONLY, &handle);

    if (err != ESP_OK) {
        retain->main_seq_step = PLC_STEP_STOPPED;
        retain->trip_latched = false;
        retain->starts = 0;
        return err;
    }

    size_t required_size = sizeof(*retain);
    err = nvs_get_blob(handle, NVS_KEY_RETAIN, retain, &required_size);
    nvs_close(handle);

    if (err != ESP_OK || required_size != sizeof(*retain)) {
        retain->main_seq_step = PLC_STEP_STOPPED;
        retain->trip_latched = false;
        retain->starts = 0;
        return err;
    }

    return ESP_OK;
}

esp_err_t plc_runtime_save_retain(const archive_runtime_t *runtime)
{
    nvs_handle_t handle;
    esp_err_t err = nvs_open(NVS_NAMESPACE, NVS_READWRITE, &handle);

    if (err != ESP_OK) {
        return err;
    }

    err = nvs_set_blob(handle, NVS_KEY_RETAIN, &runtime->retain, sizeof(runtime->retain));
    if (err == ESP_OK) {
        err = nvs_commit(handle);
    }

    nvs_close(handle);
    return err;
}

esp_err_t plc_runtime_init(archive_runtime_t *runtime)
{
    if (runtime == NULL) {
        return ESP_ERR_INVALID_ARG;
    }

    memset(runtime, 0, sizeof(*runtime));
    set_safe_outputs(runtime);

    runtime->io_config.analog_min_raw = 0.0f;
    runtime->io_config.analog_max_raw = 4095.0f;
    runtime->io_config.analog_min_eng = 0.0f;
    runtime->io_config.analog_max_eng = 100.0f;

    esp_err_t err = load_retain(&runtime->retain);

    if (runtime->retain.main_seq_step == PLC_STEP_RUNNING ||
        runtime->retain.main_seq_step == PLC_STEP_STARTING) {
        runtime->retain.main_seq_step = PLC_STEP_STOPPED;
        runtime->retain.trip_latched = true;
        plc_runtime_save_retain(runtime);
    }

    if (err != ESP_OK) {
        ESP_LOGW(TAG, "Retain load defaulted: %s", esp_err_to_name(err));
    }

    return ESP_OK;
}

void plc_runtime_scan(archive_runtime_t *runtime, const archive_inputs_t *inputs)
{
    if (runtime == NULL || inputs == NULL) {
        return;
    }

    runtime->start_pulse = inputs->start_pb && !runtime->start_prev;
    runtime->stop_pulse = inputs->stop_pb && !runtime->stop_prev;
    runtime->reset_pulse = inputs->reset_pb && !runtime->reset_prev;

    runtime->start_prev = inputs->start_pb;
    runtime->stop_prev = inputs->stop_pb;
    runtime->reset_prev = inputs->reset_pb;

    runtime->process_value = archive_io_scale_analog(inputs->adc_raw, &runtime->io_config);

    runtime->active_trip =
        !inputs->estop_ok ||
        !inputs->drive_ready ||
        inputs->drive_fault;

    if (runtime->active_trip) {
        runtime->retain.trip_latched = true;
        runtime->retain.main_seq_step = PLC_STEP_FAULTED;
    }

    if (runtime->reset_pulse && !runtime->active_trip) {
        runtime->retain.trip_latched = false;
        if (runtime->retain.main_seq_step == PLC_STEP_FAULTED) {
            runtime->retain.main_seq_step = PLC_STEP_STOPPED;
        }
        plc_runtime_save_retain(runtime);
    }

    runtime->permissive_ok =
        inputs->estop_ok &&
        inputs->drive_ready &&
        !inputs->drive_fault &&
        !runtime->retain.trip_latched;

    switch ((plc_step_t)runtime->retain.main_seq_step) {
    case PLC_STEP_STOPPED:
        set_safe_outputs(runtime);

        if (runtime->start_pulse && runtime->permissive_ok) {
            runtime->retain.main_seq_step = PLC_STEP_PRECHECK;
            plc_runtime_save_retain(runtime);
        }
        break;

    case PLC_STEP_PRECHECK:
        set_safe_outputs(runtime);

        if (!runtime->permissive_ok) {
            runtime->retain.main_seq_step = PLC_STEP_FAULTED;
            plc_runtime_save_retain(runtime);
        } else {
            runtime->retain.main_seq_step = PLC_STEP_STARTING;
            plc_runtime_save_retain(runtime);
        }
        break;

    case PLC_STEP_STARTING:
        runtime->outputs.drive_run_cmd = runtime->permissive_ok && !runtime->stop_pulse;
        runtime->outputs.run_light = runtime->outputs.drive_run_cmd;
        runtime->outputs.alarm_light = false;

        if (inputs->drive_running) {
            runtime->retain.starts++;
            runtime->retain.main_seq_step = PLC_STEP_RUNNING;
            plc_runtime_save_retain(runtime);
        } else if (!runtime->outputs.drive_run_cmd) {
            runtime->retain.main_seq_step = PLC_STEP_STOPPED;
            plc_runtime_save_retain(runtime);
        }
        break;

    case PLC_STEP_RUNNING:
        runtime->outputs.drive_run_cmd = runtime->permissive_ok && !runtime->stop_pulse;
        runtime->outputs.run_light = runtime->outputs.drive_run_cmd;
        runtime->outputs.alarm_light = false;

        if (!runtime->outputs.drive_run_cmd) {
            runtime->retain.main_seq_step = PLC_STEP_STOPPING;
            plc_runtime_save_retain(runtime);
        }
        break;

    case PLC_STEP_STOPPING:
        set_safe_outputs(runtime);

        if (!inputs->drive_running) {
            runtime->retain.main_seq_step = PLC_STEP_STOPPED;
            plc_runtime_save_retain(runtime);
        }
        break;

    case PLC_STEP_FAULTED:
        set_safe_outputs(runtime);
        runtime->outputs.alarm_light = true;
        break;

    default:
        set_safe_outputs(runtime);
        runtime->retain.trip_latched = true;
        runtime->retain.main_seq_step = PLC_STEP_FAULTED;
        plc_runtime_save_retain(runtime);
        break;
    }
}

File: simulation-projects/archive-pump-controller-sim/src/plcRuntime.mjs
export const PLC_STEP = Object.freeze({
  STOPPED: 0,
  PRECHECK: 10,
  STARTING: 20,
  RUNNING: 30,
  STOPPING: 40,
  FAULTED: 90
});

export const DEFAULT_REMOTE_COMMAND_EXPIRY_MS = 1000;

export function defaultInputs(overrides = {}) {
  return {
    estopOk: true,
    startPb: false,
    stopPb: false,
    resetPb: false,
    driveReady: true,
    driveRunning: false,
    driveFault: false,
    adcRaw: 0,
    remoteCommand: null,
    remoteCommandAgeMs: 0,
    commsHealthy: true,
    ...overrides
  };
}

export function defaultRuntime(retainOverrides = {}) {
  return {
    startPrev: false,
    stopPrev: false,
    resetPrev: false,
    startPulse: false,
    stopPulse: false,
    resetPulse: false,
    permissiveOk: false,
    activeTrip: false,
    processValue: 0,
    retain: {
      mainSeqStep: PLC_STEP.STOPPED,
      tripLatched: false,
      starts: 0,
      ...retainOverrides
    },
    outputs: safeOutputs(),
    ioConfig: {
      analogMinRaw: 0,
      analogMaxRaw: 4095,
      analogMinEng: 0,
      analogMaxEng: 100
    }
  };
}

export function initRuntime(retainOverrides = {}) {
  const runtime = defaultRuntime(retainOverrides);

  if (
    runtime.retain.mainSeqStep === PLC_STEP.RUNNING ||
    runtime.retain.mainSeqStep === PLC_STEP.STARTING
  ) {
    runtime.retain.mainSeqStep = PLC_STEP.STOPPED;
    runtime.retain.tripLatched = true;
  }

  return runtime;
}

export function scan(runtime, rawInputs = {}) {
  const inputs = defaultInputs(rawInputs);
  const remoteStart = validRemoteCommand(inputs, "start");
  const remoteStop = validRemoteCommand(inputs, "stop");
  const startRequest = inputs.startPb || remoteStart;
  const stopRequest = inputs.stopPb || remoteStop;

  runtime.startPulse = startRequest && !runtime.startPrev;
  runtime.stopPulse = stopRequest && !runtime.stopPrev;
  runtime.resetPulse = inputs.resetPb && !runtime.resetPrev;

  runtime.startPrev = startRequest;
  runtime.stopPrev = stopRequest;
  runtime.resetPrev = inputs.resetPb;

  runtime.processValue = scaleAnalog(inputs.adcRaw, runtime.ioConfig);
  runtime.activeTrip = !inputs.estopOk || !inputs.driveReady || inputs.driveFault;

  if (runtime.activeTrip) {
    runtime.retain.tripLatched = true;
    runtime.retain.mainSeqStep = PLC_STEP.FAULTED;
  }

  if (runtime.resetPulse && !runtime.activeTrip) {
    runtime.retain.tripLatched = false;
    if (runtime.retain.mainSeqStep === PLC_STEP.FAULTED) {
      runtime.retain.mainSeqStep = PLC_STEP.STOPPED;
    }
  }

  runtime.permissiveOk =
    inputs.estopOk &&
    inputs.driveReady &&
    !inputs.driveFault &&
    !runtime.retain.tripLatched;

  switch (runtime.retain.mainSeqStep) {
    case PLC_STEP.STOPPED:
      runtime.outputs = safeOutputs();
      if (runtime.startPulse && runtime.permissiveOk) {
        runtime.retain.mainSeqStep = PLC_STEP.PRECHECK;
      }
      break;

    case PLC_STEP.PRECHECK:
      runtime.outputs = safeOutputs();
      if (!runtime.permissiveOk) {
        runtime.retain.mainSeqStep = PLC_STEP.FAULTED;
      } else {
        runtime.retain.mainSeqStep = PLC_STEP.STARTING;
      }
      break;

    case PLC_STEP.STARTING:
      runtime.outputs.driveRunCmd = runtime.permissiveOk && !runtime.stopPulse;
      runtime.outputs.runLight = runtime.outputs.driveRunCmd;
      runtime.outputs.alarmLight = false;
      if (inputs.driveRunning) {
        runtime.retain.starts += 1;
        runtime.retain.mainSeqStep = PLC_STEP.RUNNING;
      } else if (!runtime.outputs.driveRunCmd) {
        runtime.retain.mainSeqStep = PLC_STEP.STOPPED;
      }
      break;

    case PLC_STEP.RUNNING:
      runtime.outputs.driveRunCmd = runtime.permissiveOk && !runtime.stopPulse;
      runtime.outputs.runLight = runtime.outputs.driveRunCmd;
      runtime.outputs.alarmLight = false;
      if (!runtime.outputs.driveRunCmd) {
        runtime.retain.mainSeqStep = PLC_STEP.STOPPING;
      }
      break;

    case PLC_STEP.STOPPING:
      runtime.outputs = safeOutputs();
      if (!inputs.driveRunning) {
        runtime.retain.mainSeqStep = PLC_STEP.STOPPED;
      }
      break;

    case PLC_STEP.FAULTED:
      runtime.outputs = { ...safeOutputs(), alarmLight: true };
      break;

    default:
      runtime.outputs = safeOutputs();
      runtime.retain.tripLatched = true;
      runtime.retain.mainSeqStep = PLC_STEP.FAULTED;
      break;
  }

  return runtime;
}

export function scaleAnalog(raw, config) {
  if (!config || config.analogMaxRaw <= config.analogMinRaw) {
    return 0;
  }

  const normalized = clamp(
    (raw - config.analogMinRaw) / (config.analogMaxRaw - config.analogMinRaw),
    0,
    1
  );

  return config.analogMinEng + normalized * (config.analogMaxEng - config.analogMinEng);
}

export function safeOutputs() {
  return {
    driveRunCmd: false,
    alarmLight: false,
    runLight: false
  };
}

function validRemoteCommand(inputs, command) {
  return (
    inputs.commsHealthy &&
    inputs.remoteCommand === command &&
    inputs.remoteCommandAgeMs <= DEFAULT_REMOTE_COMMAND_EXPIRY_MS
  );
}

function clamp(value, min, max) {
  return Math.min(Math.max(value, min), max);
}
